extensions/context/src/main/java/io/opentelemetry/context/Context.java

```diff
@@ -0,0 +1,302 @@
+    return storage().rootContext();
+  }
+
+  /** Return the context associated with the current scope. */
+  public static Context current() {
+    return storage().current();
+  }
+
+  /**
+   * Returns the default {@link ContextStorage} used to attach {@link Context}s to scopes of
+   * execution. Should only be used when defining your own {@link ContextStorage} in case you want
+   * to delegate functionality to the default implementation.
+   */
+  public static ContextStorage threadLocalStorage() {
+    return ThreadLocalContextStorage.INSTANCE;
+  }
+
+  static ContextStorage storage() {
+    return LazyStorage.storage;
+  }
+
+  private final PersistentHashArrayMappedTrie.Node<ContextKey<?>, Object> entries;
+
+  private Context(PersistentHashArrayMappedTrie.Node<ContextKey<?>, Object> entries) {
+    this.entries = entries;
+  }
+
+  Context() {
+    entries = null;
+  }
+
+  /**
+   * Returns the value stored in this {@link Context} for the given {@link ContextKey}, or {@code
+   * null} if there is no value for the key in this context.
+   */
+  @Nullable
+  public <V> V getValue(ContextKey<V> key) {
+    // Because withValue enforces the value for a key is its type, this is always safe.
+    @SuppressWarnings("unchecked")
+    V value = (V) PersistentHashArrayMappedTrie.get(entries, key);
+    return value;
+  }
+
+  /**
+   * Returns a new context with the given key value set.
+   *
+   * <pre>{@code
+   * Context withCredential = Context.current().withValue(CRED_KEY, cred);
+   * withCredential.run(new Runnable() {
+   *   public void run() {
+   *      readUserRecords(userId, CRED_KEY.get());
+   *   }
+   * });
+   * }</pre>
+   *
+   * <p>Note that multiple calls to {@link #withValue(ContextKey, Object)} can be chained together.
+   * That is,
+   *
+   * <pre>
+   * context.withValues(K1, V1, K2, V2);
+   * // is the same as
+   * context.withValue(K1, V1).withValue(K2, V2);
+   * </pre>
+   *
+   * <p>Nonetheless, {@link Context} should not be treated like a general purpose map with a large
+   * number of keys and values — combine multiple related items together into a single key instead
+   * of separating them. But if the items are unrelated, have separate keys for them.
+   */
+  public <V> Context withValue(ContextKey<V> k1, V v1) {
+    PersistentHashArrayMappedTrie.Node<ContextKey<?>, Object> newEntries =
+        PersistentHashArrayMappedTrie.put(entries, k1, v1);
+    return new Context(newEntries);
+  }
+
+  /** Returns a new context with the given key value set. */
+  public <V1, V2> Context withValues(ContextKey<V1> k1, V1 v1, ContextKey<V2> k2, V2 v2) {
+    PersistentHashArrayMappedTrie.Node<ContextKey<?>, Object> newEntries =
+        PersistentHashArrayMappedTrie.put(entries, k1, v1);
+    newEntries = PersistentHashArrayMappedTrie.put(newEntries, k2, v2);
+    return new Context(newEntries);
+  }
+
+  /** Returns a new context with the given key value set. */
+  public <V1, V2, V3> Context withValues(
```

Comment:
Do we need all the overrides at this point? I'd like to figure out how minimal we can make this API for starters.